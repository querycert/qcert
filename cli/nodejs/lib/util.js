/*
 * Licensed under the Apache License; Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing; software
 * distributed under the License is distributed on an "AS IS" BASIS;
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND; either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* Stringify */
function replacer(key, value) {
  if (typeof value === 'bigint') {
    return value.toString() + 'n';
  }
  return value;
}

function stringified(data) {
  if (typeof data === 'bigint') {
    return data.toString() + 'n';
  }
  return JSON.stringify(data, replacer);
}

/* Generic Node module require from string */
function requireFromString(src, filename) {
  const Module = module.constructor;
  const m = new Module();
  m._compile(src, filename);
  return m.exports;
}

module.exports = { stringified, requireFromString };
