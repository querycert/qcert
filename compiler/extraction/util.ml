(*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *)

(* This module contains a few basic utilities *)

(* Qcert Exception *)

exception Qcert_Error of string

(* this can't go in Logger, since that creates a circular dependency *)
type nraenv_logger_token_type = string
type nnrc_logger_token_type = string
type nnrs_imp_expr_logger_token_type = string
type nnrs_imp_stmt_logger_token_type = string
type nnrs_imp_logger_token_type = string
type dnnrc_logger_token_type = string

(**************)
(* Data types *)
(**************)

(* Data type conversions between Coq and OCaml *)

let string_of_char_list l =
  let b = Bytes.create (List.length l) in
  let i = ref 0 in
  List.iter (fun c -> Bytes.set b !i c; incr i) l;
  Bytes.to_string b

let char_list_of_string s =
  let rec exp i l =
    if i < 0 then l else exp (i - 1) (s.[i] :: l) in
  exp (String.length s - 1) []

let string = string_of_char_list

let flat_map_string f s =
  let sl = ref [] in
  String.iter (fun c -> sl := (f c) :: !sl) s;
  let sl' = List.rev !sl in
  String.concat "" sl'

(* coq Z's are now replaced by native OCaml ints, but here is the way to get things back to coq Z's:

open BinNums

let rec coq_nat_of_pos i =
  if i = 0 then Datatypes.O else Datatypes.S (coq_nat_of_pos (i-1))

let coq_positive_of_pos i =
  BinPos.Pos.of_nat (coq_nat_of_pos i)

let coq_Z_of_int i =
  if (i = 0) then Z0
  else if (i < 0)
  then (Zneg (coq_positive_of_pos (-i)))
  else (Zpos (coq_positive_of_pos i))
 *)

let coq_Z_of_int i = i


(*******)
(* I/O *)
(*******)

(* Temporarily disabled -- JS
   let os_newline = if Sys.win32 then "\r\n" else "\n" *)
let os_newline = "\n"

let string_of_file file =
  try
    let inchan = open_in_bin file in
    let len = in_channel_length inchan in
    let buf = Buffer.create len in
    Buffer.add_channel buf inchan len;
    close_in inchan;
    Buffer.contents buf
  with
    Sys_error err ->
      Printf.eprintf
        "Could not read the file %s, got error Sys_error %s\n@?"
        file
        err;
      raise(Sys_error err)

(* File print *)

let make_file fout scomp =
  let oc = open_out fout in
  begin
    output_string oc scomp;
    close_out oc
  end

(* Make up target file name *)

let target_f dir f =
  match dir with
  | None -> f
  | Some d ->
    Filename.concat d (Filename.basename f)

let outname f suff = f ^ suff


(**********************************)
(* Support for Enhanced operators *)
(**********************************)

let qcert_string_of_float f =
  let ocaml_string = string_of_float f in
  let last_char = ocaml_string.[(String.length ocaml_string)-1] in
  match last_char with
  | '.' -> ocaml_string ^ "0"
  | _ -> ocaml_string

let string_of_enhanced_float f = char_list_of_string (string_of_float f)
let string_of_enhanced_string s = char_list_of_string ("S\"" ^ s ^ "\"")
	
(**********************************)
(* Timing function for CompStat   *)
(**********************************)

let time f x =
  let start = Sys.time() in
  let v = f x in
  let stop = Sys.time() in
  let t = string_of_float (stop -. start) in
  (char_list_of_string t, v)

(* String manipulation *)    

let string_after s n = String.sub s n (String.length s - n)

let re_match re s pos =
  if pos >= String.length s then raise Not_found
  else
    let rec pos_match re s pos_re pos_s =
      if pos_re >= String.length re
      then true
      else
	try
	  if re.[pos_re] = s.[pos_s]
	  then pos_match re s (pos_re+1) (pos_s+1)
	  else false
	with
	| _ -> false
    in
    pos_match re s 0 pos
  
let search_forward re s pos =
  if re = "" then raise (Invalid_argument "Matching string should not be empty")
  else
    let rec find start =
      if re_match re s start
      then start
      else find (start+1)
    in
    find pos

let opt_search_forward re s pos =
  try Some(search_forward re s pos) with Not_found -> None

let global_replace const_expr repl text =
  let rec replace accu start last_was_empty =
    let startpos = if last_was_empty then start + 1 else start in
    if startpos > String.length text then
      string_after text start :: accu
    else
      match opt_search_forward const_expr text startpos with
      | None ->
          string_after text start :: accu
      | Some pos ->
          let end_pos = pos + String.length const_expr in
          let repl_text = repl in
          replace (repl_text :: String.sub text start (pos-start) :: accu)
                  end_pos (end_pos = pos)
  in
  String.concat "" (List.rev (replace [] 0 false))

  
