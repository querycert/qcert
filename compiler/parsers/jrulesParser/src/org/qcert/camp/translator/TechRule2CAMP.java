/**
 * Copyright (C) 2017 Joshua Auerbach 
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.qcert.camp.translator;

import java.io.StringReader;

import org.qcert.camp.pattern.CampPattern;
import org.qcert.javasvc.Command;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

/**
 * Implement Java Service command for converting a technical rule (ie, an "ARL file") to CAMP
 */
public class TechRule2CAMP implements Command {
	@Override
	public String invoke(String arg) {
		try {
			StringReader rdr = new StringReader(arg);
			JsonObject input = new JsonParser().parse(rdr).getAsJsonObject();
			JsonElement schema = input.get("schema");
			rdr = new StringReader(input.get("source").getAsString());
			ODMFrontEnd frontEnd = new ODMFrontEnd(rdr, null, schema);
			SemRule2CAMP semRule2Camp = new SemRule2CAMP(frontEnd.getFactory());
			CampPattern translated = semRule2Camp.translate(frontEnd.getRule()).convertToPattern();
			return translated.emit();
		} catch (Exception e) {
			return "ERROR: " + e.getMessage() + ", while parsing " + arg;
		}
	}
}
