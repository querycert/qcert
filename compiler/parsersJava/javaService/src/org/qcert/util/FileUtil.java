/**
 * Copyright (C) 2016 Joshua Auerbach 
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.qcert.util;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;

/**
 * Trivial file utilities
 */
public class FileUtil {
	/**
	 * Read text from a file into a String
	 * @param srcFile the input file
	 */
	public static String readFile(File srcFile) throws IOException {
		StringWriter swtr = new StringWriter();
		PrintWriter wtr = new PrintWriter(swtr);
		FileReader frdr = new FileReader(srcFile);
		BufferedReader rdr = new BufferedReader(frdr);
		String line = rdr.readLine();
		while (line != null) {
			wtr.println(line);
			line = rdr.readLine();
		}
		rdr.close();
		wtr.close();
		return swtr.toString();
	}

	/**
	 * Write text to a file
	 * @param file the file
	 * @param text the text to write
	 */
	public static void writeFile(File file, String text) throws IOException {
		PrintWriter wtr = new PrintWriter(new FileWriter(file));
		wtr.println(text);
		wtr.close();
	}

}
