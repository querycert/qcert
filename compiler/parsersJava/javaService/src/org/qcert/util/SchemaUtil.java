/**
 * Copyright (C) 2017 Joshua Auerbach 
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.qcert.util;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

/**
 * Contains utilities to convert JSON-formatted schemas into a more convenient form.  We use a simple inheritance of type representations.
 * <p>A <b>Type</b> is one of the following (the general case represented by a marker interface)
 * <ul><li>a <b>PrimitiveType</b> (containing a string type name), 
 * <li>a <b>ListType</b> (containing an element Type),
 * <li>an <b>ObjectType</b>, which has a brand and a attributes.  The attributes are represented as an ordered map from names 
 *   to <b>Type</b> objects.
 * <li>A schema is just a map from brands to <b>ObjectTypes</b> with those brands.  
 *   
 */
public class SchemaUtil {
	private SchemaUtil() {}
	
	/**
	 * Utility to derive the convenient schema form from various JSON encodings.  We support several formats.
	 * <ul><li>The so-called "I/O" file format, in which there is a "schema" member.
	 * <li>A JSON object consisting only of the schema (the same as the "schema" member of the I/O file but as a simple JSON object in a file
	 *   by itself)
	 * <li>A JSON object consisting only of the model portion of the schema.
	 * <li>A simplified form of the model portion of the schema in which the brandTypes and typeDefs are collapsed into a single
	 *   array of JSON objects, with each object having a "brand" and "typeDef" member (unifying on the "type name" which is not really
	 *   needed here)
	 * </ul>
	 * @param encoded a JsonElement encoding a schema in one of the supported formats
	 * @return a Map from brands to ObjectTypes defining the attributes associated with those brands
	 */
	public static Map<String, ObjectType> getSchema(JsonElement encoded) {
		if (encoded.isJsonObject()) {
			JsonObject objForm = encoded.getAsJsonObject();
			/* Deal with possibly different levels of wrapping */
			if (objForm.has("schema"))
				objForm = objForm.get("schema").getAsJsonObject();
			if (objForm.has("model")) // now obsolete but may persist in some corners
				objForm = objForm.get("model").getAsJsonObject();
			if (objForm.has("brandTypes") && objForm.has("typeDefs")) {
				JsonArray brandTypes = objForm.get("brandTypes").getAsJsonArray();
				JsonArray typeDefs = objForm.get("typeDefs").getAsJsonArray();
				return getSchemaFromBrandsAndTypedefs(brandTypes, typeDefs);
			} else if (objForm.has("types"))
				return getSchemaFromTypes(objForm.get("types").getAsJsonArray());
			else
				throw new IllegalArgumentException("Uninterpretable JSON element");
		} else if (encoded.isJsonArray())
			return getSchema(encoded.getAsJsonArray());
		else
			throw new IllegalArgumentException("Uninterpretable JSON element");
	}
	
	public static Map<String, ObjectType> getGlobalsFromSchema(JsonObject schema) {
		/* Deal with possibly different levels of wrapping */
		if (schema.has("schema"))
			schema = schema.get("schema").getAsJsonObject();
		if (schema.has("model")) // now obsolete but may persist in some corners
			schema = schema.get("model").getAsJsonObject();
		Map<String, ObjectType> ans = new HashMap<>();
		if (schema.has("globals")) {
			schema = schema.get("globals").getAsJsonObject();
			if (schema.size() == 1 && schema.has("WORLD"))
				// We heuristically assume the globals are uninteresting in the "one WORLD" case.  Use getSchema if the important
				// information is contained in the branded types.
				return ans;
			for (Entry<String, JsonElement> member : schema.entrySet()) {
				String typeName = member.getKey();
				JsonObject typeDef = member.getValue().getAsJsonObject().get("type").getAsJsonObject().get("$coll").getAsJsonObject();
				ans.put(typeName, makeObjectType(typeName, typeDef));
			}
		}
		return ans;
	}

	/**
	 * Utility to derive a convenient schema from two arrays, one pairing brands with type names, and the other pairing type names
	 *   with typedefs.  This is one form in which schemas are encoded (e.g. in the I/O file)
	 * @param brandTypes the array of pairs linking brands to type names
	 * @param typeDefs the array of pairs linking type names to type defs
	 * @return the convenient form of the schema
	 */
	public static Map<String, ObjectType> getSchemaFromBrandsAndTypedefs(JsonArray brandTypes, JsonArray typeDefs) {
		Map<String, String> brandMap = new HashMap<>();
		for (JsonElement pair : brandTypes) {
			JsonObject brandPair = pair.getAsJsonObject();
			brandMap.put(brandPair.get("typeName").getAsString(), brandPair.get("brand").getAsString());
		}
		Map<String, ObjectType> result = new HashMap<>();
		for (JsonElement pair : typeDefs) {
			JsonObject typePair = pair.getAsJsonObject();
			String brand = brandMap.get(typePair.get("typeName").getAsString());
			result.put(brand, makeObjectType(brand, typePair.get("typeDef").getAsJsonObject()));
		}
		return result;
	}

	/**
	 * Utility to derive a convenient schema from a single array which pairs brand names with typedefs.  Unlike getSchemaFromBrandsAndTypedefs, this encoding can't handle
	 *   embedded objects (other than collections).  Also, it is not currently the standard for I/O files.  However it is more straightforward to produce in many cases.
	 * @param types the JsonArray of types as brand/typeDef pairs
	 * @return the simplified schema
	 */
	public static Map<String, ObjectType> getSchemaFromTypes(JsonArray types) {
		Map<String, ObjectType> result = new HashMap<>();
		for (JsonElement pair : types) {
			JsonObject typePair = pair.getAsJsonObject();
			String brand = typePair.get("brand").getAsString();
			result.put(brand, makeObjectType(brand, typePair.get("typeDef").getAsJsonObject()));
		}
		return result;
	}

	/**
	 * Make an ObjectType from a brand name and typeDef
	 * @param brand the brand name
	 * @param typeDef the typeDef (as a JsonObject)
	 * @return the ObjectType
	 */
	private static ObjectType makeObjectType(String brand, JsonObject typeDef) {
		ObjectType ans = new ObjectType(brand);
		for (Entry<String, JsonElement> attr : typeDef.entrySet()) {
			ans.addAttribute(attr.getKey(), makeType(attr.getValue()));
		}
		return ans;
	}

	/**
	 * Make a Type from a JsonElement representing a Type.  
	 * TODO support for embedded objects (other than collections) is incomplete.
	 * @param type the JsonElement to be made into a Type
	 * @param brandMap the map from type names to brands in the event of an embedded objectl
	 * @return the Type
	 */
	private static Type makeType(JsonElement type) {
		if (type.isJsonPrimitive() && type.getAsJsonPrimitive().isString()) {
			return new PrimitiveType(type.getAsJsonPrimitive().getAsString());
		}
		// Other than primitive types and object type references we only support the $coll convention at this time
		JsonElement coll = type.getAsJsonObject().get("$coll");
		if (coll == null)
			throw new UnsupportedOperationException("Don't yet know how to handle encoded type " + type);
		Type element = makeType(coll);
		return new ListType(element);
	}
	
	/** List Type */
	public static class ListType implements Type {
		public final Type elementType;
		public ListType(Type elementType) {
			this.elementType = elementType;
		}
	}

	/** Object Type */
	public static class ObjectType implements Type {
		public final String brand;
		public final LinkedHashMap<String, Type> attributes = new LinkedHashMap<>();
		public ObjectType(String brand) {
			this.brand = brand;
		}
		public void addAttribute(String name, Type type) {
			attributes.put(name, type);
		}
	}
	
	/** Primitive Type */
	public static class PrimitiveType implements Type {
		public final String typeName;
		public PrimitiveType(String typeName) {
			this.typeName = typeName;
		}
	}
	
	/** Marker interface for all types */
	public interface Type {}
}
