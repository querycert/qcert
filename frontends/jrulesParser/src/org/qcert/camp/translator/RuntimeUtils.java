/**
 * (c) Copyright IBM Corp. 2015-2017
 * Copyright (C) 2017 Joshua Auerbach 
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.qcert.camp.translator;

import java.util.Collections;
import java.util.List;

/** Some utility functions used as support at runtime by code that the jrules2coq translator generates */
public class RuntimeUtils {
	/**
	 * Utility function called from the ODM runtime to sort and format a list of strings in a consistent way
	 * @param ls the list
	 * @return the formatted result
	 */
	public static String stringListToString(List<String> ls) {
		Collections.sort(ls);
		return ls.toString();
	}
}
