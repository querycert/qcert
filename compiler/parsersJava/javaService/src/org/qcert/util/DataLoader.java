/**
 * Copyright (C) 2016 Joshua Auerbach 
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.qcert.util;

import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVParser;
import org.apache.commons.csv.CSVRecord;
import org.qcert.util.SchemaUtil.ListType;
import org.qcert.util.SchemaUtil.ObjectType;
import org.qcert.util.SchemaUtil.PrimitiveType;
import org.qcert.util.SchemaUtil.Type;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.JsonPrimitive;
import com.google.gson.internal.Streams;
import com.google.gson.stream.JsonWriter;

/**
 * Generates the standard qcert input format from one or more "CSV" files and a qcert schema.
 * 
 * <p>There are two public methods, a <b>loadData</b> method, which is flexible enough to handle all sorts of file naming conventions
 *   and "CSV" formats, and a <b>main</b> method, which can be invoked from the command line but requires a strict file naming
 *   convention and handles only strict the RFC 4180 CSV interpretation, with a required header line in each file.
 *    
 * <p>Most generally, the input consists of
 * <ol><li>a set of CSV files and their associated internal type ("table") names.  On the command line, the association is strict:
 *   the file name must be the table name plus the suffix <b>.csv</b> and be found in a specific directory.  For <b>loadData</b>
 *   the association is given explicitly in a Map argument.
 * <li>a CSV format designation (all formats supported by Apache commons-csv, but a header must be available somehow, 
 *  either from the format or from each file).  On the command line, the format is not specified and must be RFC 4180 with in-file
 *  headers (the delimiter may optionally be set to something other than comma).
 *  <li>a qcert schema.  On the command line, this is given as a file path.  For <b>loadData</b> it is given as a <b>JsonElement</b>.
 *  </ol>
 * <p>Each CSV file should contain just data for its corresponding internal type.  That type may be either a brand in 
 *   the schema's <b>brandTypes</b> or a partition name in the schema's <b>globals</b>, depending on the memory
 *  layout intended.
 * <p>The schema should be in one of qcert standard JSON formats supported by <b>SchemaUtil</b>.  
 *   The two different ways that the set of "table" names can match names in the schema gives rise to a different memory layout.
 * <ol><li>If the table name matches a brand in the <b>brandTypes</b>, the type definition is taken from <b>typeDefs</b>
 *  and the input of that type will go in the partition called "WORLD", marked by its type.  
 *  This is a heuristic but works for present cases because, depending on the source language, we either assume this case or the
 *   next one.
 *  <li>If the table name matches a member of <b>globals</b>, the type definition is the <b>type</b> of that member
 *  and input of that type will go in a homogeneous memory partition named for the type.
 *  </ol>
 * In either case, the names of the attributes in the type definition are matched against header names in the CSV file.  A column in 
 * the CSV file that does not match an attribute name is an error.
 * <p>For this to work, all types in the schema must be primitive (numbers, booleans strings, and dates).  Embedded list and
 * object types are not handled.
 * <p>Currently, if the type is "date" (ignoring case), the field is formatted as a date JSON object with members "year", "month" and "day".
 * This is a good convention for SQL but other formats probably need to be investigated.
 */
public class DataLoader {
	private DataLoader() {}
	
	/** Primary entry point to this service.  The 'main' method is also useful but can be bypassed to call this instead
	 *   when files are named in non-standard ways or to use a CSV format understood by Apache commons-csv but other than strict RFC 4180.
	 * @param tableMap a map from "table names" (type names matching CSV files) to the corresponding CSV file contents (as Strings)
	 * @param jsonSchema the JSON format schema in one of several possible variations, as a JsonElement (array or object)
	 * @param format the CSV format to use (if the format does not include a header we assume that the first line of 
	 *   each file constitutes a header
	 * @return a JsonObject representing the loaded data
	 * @throws Exception
	 */
	public static JsonObject loadData(Map<String, String> tableMap, JsonElement jsonSchema, CSVFormat format) throws Exception {
		if (format.getHeader() == null)
			format = format.withHeader();
		Map<String, ObjectType> types = SchemaUtil.getSchema(jsonSchema);
		boolean partitioned = types.size() == 0;
		if (partitioned) {
			if (jsonSchema.isJsonObject())
				types = SchemaUtil.getGlobalsFromSchema(jsonSchema.getAsJsonObject());
			if (types.size() == 0)
				throw new IllegalArgumentException("Schema contains no useful information");
		}
			
		/* Process the data files */
		JsonObject ans = new JsonObject();
		JsonArray world = partitioned ? null : new JsonArray();
		if (!partitioned)
			ans.add("WORLD", world);
		for (Entry<String, String> filePair : tableMap.entrySet()) {
			String table = filePair.getKey();
			ObjectType def = types.get(table);
			if (def == null)
				throw new IllegalArgumentException("Type " + table + " is not in the schema");
			JsonArray thisType = process(filePair.getValue(), def, format);
			if (partitioned)
				ans.add(table, thisType);
			else {
				for (JsonElement elem : thisType) {
					JsonObject toAdd = new JsonObject();
					JsonArray brands = new JsonArray();
					brands.add(new JsonPrimitive(table));
					toAdd.add("type", brands);
					toAdd.add("data", elem);
					world.add(toAdd);
				}
			}
		}
		return ans;
	}

	/**
	 * Main program.  
	 * <p>Command line arguments are
	 * <ul><li><b>-dir &lt;path&gt;</b> (optional) the directory in which to find or create all other files, defaults to current directory
	 * <li><b>-output &lt;filename&gt;</b> (required) the name of the output file (absolute or relative to <b>-dir</b>)
	 * <li><b>-schema &lt;filename&gt;</b> (required) the name of the schema file (absolute or relative to <b>-dir</b>)
	 * <li><b>-delimiter &lt;char&gl;</b> (optional) a character to use as delimiter (defaults to comma; if the chose character is shell-sensitive, be sure to quote it)
	 * <li>all other arguments are assumed to be type names.  A file of that name with extension <b>.csv</b> must be present in <b>-dir</b>
	 * </ul>
	 */
	public static void main(String[] args) throws Exception {
		
		/* Parse the command line */
		String directory = null, schema = null, output = null; char delimiter = 0;
		List<String> tables = new ArrayList<>();
		boolean dirFlag = false, outFlag = false, schemaFlag = false, delimiterFlag = false;
		for (String arg : args) {
			boolean table = false;
			if (dirFlag)
				directory = arg;
			else if (outFlag)
				output = arg;
			else if (schemaFlag)
				schema = arg;
			else if (delimiterFlag) {
				if (arg.length() != 1)
					throw new IllegalArgumentException("Delimiters must be single character");
				else
					delimiter = arg.charAt(0);
			} else
				table = true;
			dirFlag = outFlag = schemaFlag = delimiterFlag = false;
			if (arg.charAt(0) == '-') {
				switch(arg) {
				case "-dir":
					if (directory != null)
						throw new IllegalArgumentException("Duplicate -dir");
					else
						dirFlag = true;
					continue;
				case "-output":
					if (output != null)
						throw new IllegalArgumentException("Duplicate -output");
					else
						outFlag = true;
					continue;
				case "-schema":
					if (schema != null)
						throw new IllegalArgumentException("Duplicate -schema");
					else
						schemaFlag = true;
					continue;
				case "-delimiter":
					if (delimiter != 0)
						throw new IllegalArgumentException("Duplicate -delimiter");
					else
						delimiterFlag = true;
					continue;
				default:
					throw new IllegalArgumentException("Unknown option :" + arg);
				}
			}
			if (table)
				tables.add(arg);
		}
		if (output == null)
			throw new IllegalArgumentException("Output file must be specified");
		if (schema == null)
			throw new IllegalArgumentException("Schema file must be specified");
		File outputFile = new File(output);
		File schemaFile = new File(schema);
		if (directory != null) {
			if (!outputFile.isAbsolute())
				outputFile = new File(directory, output);
			if (!schemaFile.isAbsolute())
				schemaFile = new File(directory, schema);
		}
		
		/* Canonicalize the file names, read the contents, and associate each with its corresponding table name */
		Map<String, String> tableMap = new HashMap<>();
		for (String table : tables) {
			File toRead = directory == null ? new File(table + ".csv") : new File(directory, table + ".csv");
			String contents = new String(Files.readAllBytes(Paths.get(toRead.getAbsolutePath())));
			tableMap.put(table, contents);
		}
		
		/* Parse the schema */
		JsonElement jsonSchema = new JsonParser().parse(new FileReader(schemaFile));

		/* Process, using callable service */
		CSVFormat format = CSVFormat.RFC4180;
		if (delimiter != 0)
			format = format.withDelimiter(delimiter);
		JsonObject ans = loadData(tableMap, jsonSchema, format);
		
		/* Write the result */
		FileWriter wtr = new FileWriter(outputFile);
		JsonWriter json = new JsonWriter(wtr);
		json.setLenient(true);
		json.setIndent("  ");
		Streams.write(ans, json);
		json.close();
	}

	/**
	 * Convert a primitive value of designated type to a JsonElement
	 * @param value the value to convert
	 * @param fieldType the type of the field
	 * @return a JsonElement (either a primitive or a date object)
	 */
	private static JsonElement convertPrimitiveValue(String value, Type fieldType) {
		String typeName = ((PrimitiveType) fieldType).typeName;
		switch(typeName) {
		case "String":
			return new JsonPrimitive(value);
		case "Nat":
      JsonObject ans = new JsonObject();
			/* We are a little loosy-goosy with numbers since the sources of information are often inexact */
			try {
			    ans.add("$nat",new JsonPrimitive(Integer.parseInt(value)));
			    return ans;
			} catch (NumberFormatException ig) {
			    ans.add("$nat",new JsonPrimitive(Double.parseDouble(value)));
			    return ans;
			}
		case "Float":
        return new JsonPrimitive(Double.parseDouble(value));
		case "ESqlDate":
			return formatDate(value);
		case "Bool":
			return new JsonPrimitive(value.equalsIgnoreCase("true"));
		default:
			throw new UnsupportedOperationException("Don't known how to convert primitive schema type " + typeName);
		}
	}

	/**
	 * Format a SQL-style String date into a JSON date object, which, by convention, we use for this type
	 * TODO support other kinds of dates
	 * @param stringDate the date to format
	 * @return the JSON result
	 */
	private static JsonElement formatDate(String stringDate) {
		String[] dateParts = stringDate.split("-");
		assert dateParts.length == 3;
		String[] names = {"year",  "month", "day"};
		JsonObject ans = new JsonObject();
		JsonObject ansIn = new JsonObject();
		JsonObject ansIn2 = new JsonObject();
		for (int i = 0; i < 3; i++) {
      ansIn2.add(names[i], new JsonPrimitive(Integer.parseInt(dateParts[i])));
		}
    ansIn.add("$date", ansIn2);
    ans.add("$foreign", ansIn);
		return ans;
	}

	/** Process an individual table, producing its rows in JSON form 
	 * @param data the CSV file contents as a String
	 * @param def the type definition as an ObjectType
	 * @param format the CSVFormat to use
	 * @return a JsonArray of the translation of the rows
	 * @throws Exception
	 */
	private static JsonArray process(String data, ObjectType def, CSVFormat format) throws Exception {
		JsonArray ans = new JsonArray();
		List<CSVRecord> records = CSVParser.parse(data, format).getRecords();
		for (CSVRecord record : records) {
			Map<String, String> recmap = record.toMap();
			JsonObject datum = new JsonObject();
			for (Entry<String, String> col : recmap.entrySet()) {
				datum.add(col.getKey(), processColumn(col.getKey(), col.getValue(), def));
			}
			ans.add(datum);
		}
		return ans;
	}

	/**
	 * Process an individual column of an individual row of an individual table, producing its JSON representation as a JSON object
	 * @param fieldName the name of the column
	 * @param value the raw (String) value in the CSV file
	 * @param def the ObjectType that should produce the type for each field name, allowing the value to be interpreted
	 * @return a JsonObject encoding the column of the row
	 */
	private static JsonElement processColumn(String fieldName, String value, ObjectType def) {
		if (fieldName == null)
			throw new IllegalArgumentException("No header information in CSV file");
		Type fieldType = def.attributes.get(fieldName);
		if (fieldType == null)
			throw new IllegalArgumentException("Field " + fieldName + " is not among the attributes of type " + def.brand);
		if (fieldType instanceof PrimitiveType) {
			return convertPrimitiveValue(value, fieldType);
		}
		if (fieldType instanceof ListType) {
			Type elementType = ((ListType) fieldType).elementType;
			if (elementType instanceof PrimitiveType) {
				JsonArray ans = new JsonArray();
				String[] values = value.split(",");
				for (String val : values)
					ans.add(convertPrimitiveValue(val, elementType));
				return ans;
			}
		}
		throw new UnsupportedOperationException("Can't handle embedded object types, lists thereof, or lists of lists, when loading data");
	}
}
