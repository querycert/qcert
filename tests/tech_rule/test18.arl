import java.lang.*;
import java.util.*;
import rete.*;
import tests.branding.*;
import org.qcert.camp.translator.RuntimeUtils;
import com.ibm.ia.model.*;

ruleset test18 {
  rule FindMarketers {
    when {
      C:  Client();
      Ms: aggregate { M: Marketer(clients.contains(C.id)); }
      do { ArrayList<String> {M.name}; }
    } then { 
      System.out.println("Client " + C.name + " has marketers " + RuntimeUtils.stringListToString(Ms)); 
    }
  }
}
