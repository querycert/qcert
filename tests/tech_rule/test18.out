[["Client Librettos are Us has marketers [James Do]","Client We Got Dough has marketers [James Do]","Client Croissant City has marketers [Joan Doe]","Client Tomatoe Paradise has marketers [Jill Does]","Client Have Potatoes, Will Travel has marketers [Jill Does]","Client Stillettos are Us has marketers [Joan Doe]"]]